scripts/globals/events/starlight_celebrations.lua

```diff
@@ -0,0 +1,943 @@
+
+--------------------------------
+-- Gifts for NPCs Sub-Quest --
+--------------------------------
+
+function xi.events.starlightCelebration.npcGiftsMoogleOnTrigger(player)
+    if xi.events.starlightCelebration.isStarlightEnabled ~= 0 then
+        local startedQuest = player:getLocalVar("[StarlightNPCGifts]Started")
+        local hasCompleted = player:getCharVar("[StarlightNPCGifts]Completed")
+        if startedQuest ~= 0 then
+            local delivered = player:getLocalVar("[StarlightNPCGifts]Delivered")
+            if delivered == 15 then
+                player:startEvent(32702)
+            else
+                player:startEvent(32701, delivered)
+            end
+        else
+            if hasCompleted == VanadielDayOfTheWeek() then
+                player:startEvent(32700)
+            else
+                player:startEvent(32703)
+            end
+        end
+    end
+end
+
+function xi.events.starlightCelebration.npcGiftsMoogleOnFinish(player, id, csid, option)
+    if csid == 32703 and option == 1 then
+        player:setLocalVar("[StarlightNPCGifts]Started", 1)
+        player:setLocalVar("[StarlightNPCGifts]Delivered", 0)
+        player:setLocalVar("[StarlightNPCGifts]Npc1", 0)
+        player:setLocalVar("[StarlightNPCGifts]Npc2", 0)
+        player:setLocalVar("[StarlightNPCGifts]Npc3", 0)
+        player:setLocalVar("[StarlightNPCGifts]Npc4", 0)
+    elseif csid == 32702 then
+        local hasToken = player:hasKeyItem(xi.keyItem.SNOW_THEMED_GIFT_TOKEN)
+        local gift_table =
+        {
+            5622,   -- Candy Cane
+            5621,   -- Candy Ring
+        }
+        local fireworks_table =
+        {
+            4215, -- Popstar
+            4216, -- Brilliant Snow
+            4217, -- Sparkling Hand
+            4218, -- Air Rider
+            4167, -- Cracker
+            4168, -- Twinkle Shower
+            4169, -- Little Comet
+        }
+        if hasToken == true then
+            npcUtil.giveItem(player, gift_table[math.random(1, 2)])
+        else
+            player:addKeyItem(xi.keyItem.SNOW_THEMED_GIFT_TOKEN)
+            player:messageSpecial(id.text.KEYITEM_OBTAINED, xi.keyItem.SNOW_THEMED_GIFT_TOKEN)
+        end
+        npcUtil.giveItem(player, { { fireworks_table[math.random( 1, 7 )], 10 } })
+        player:setLocalVar("[StarlightNPCGifts]Started", 0)
+        player:setCharVar("[StarlightNPCGifts]Completed", VanadielDayOfTheWeek())
+    end
+end
+
+function xi.events.starlightCelebration.npcGiftsNPC1(player, eventid)
+    if eventid == 1 then
+        if player:getLocalVar("[StarlightNPCGifts]Started") == 1 then
+            if player:getLocalVar("[StarlightNPCGifts]Npc1") ~= 1 then
+                local npc2 = player:getLocalVar("[StarlightNPCGifts]Npc2")
+                local npc3 = player:getLocalVar("[StarlightNPCGifts]Npc3")
+                local npc4 = player:getLocalVar("[StarlightNPCGifts]Npc4")
+                local delivered = 0
+                if npc2 == 1 then
```

Comment:
These just seem to be binary bits, you could concat npc4..npc3..npc2..npc1 and set delivered to tonumber(string, 2)

Reply:
Aye, that makes more sense!

Reply:
Appreciate all the feedback, this bit in particular saved quite a few lines of code. Still getting use to using binary, hex etc! Merging shortly!